### PR TITLE
Fix UserData metadata so piping occurs correctly

### PR DESCRIPTION
**Pull request by Sandido**
## Description

The new UserData parameter was added [here](https://github.com/Azure/azure-powershell/pull/16187). The metadata tag ValueFromPipeline was used incorrectly, and ValueFromPipelineByPropertyName should have been used to ensure the correct values are piped into the UserData parameter.

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/1099
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by msJinLei** [comment]
As you code is a bug fix and user data is not released for too long. We have chance to fix it now rather than wait until next breaking change release.
To include it we need confirm from PM @dcaro.  We need to check telemetry to evaluate the effect to customers.
Please sent email to @dingmeng-xue and @dcaro.
However since they are all in vacation, we may not receive response until after the holidays.
The code complete date is **PST 6PM 12/27**, Please submit the PR earlier.

We can release your PR as OOB if it is urgent and we receive confirmation from PM.

File: tools/StaticAnalysis/Exceptions/Az.Compute/BreakingChangeIssues.csv
Comment (msJinLei) [inline tools/StaticAnalysis/Exceptions/Az.Compute/BreakingChangeIssues.csv:8]: This line should be in SignatureIssues.csv

**Review by msJinLei — CHANGES_REQUESTED** [review]
```
  D:\a\1\s\artifacts//StaticAnalysisResults\BreakingChangeIssues.csv Errors
  "AssemblyFileName","ClassName","Target","Severity","ProblemId","Description","Remediation"
  "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'DefaultParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'DefaultParamSet' back to cmdlet 'Get-AzVM'."
  "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'GetVirtualMachineInResourceGroupParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'GetVirtualMachineInResourceGroupParamSet' back to cmdlet 'Get-AzVM'."
  "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'ListLocationVirtualMachinesParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'ListLocationVirtualMachinesParamSet' back to cmdlet 'Get-AzVM'."
  "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'ListNextLinkVirtualMachinesParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'ListNextLinkVirtualMachinesParamSet' back to cmdlet 'Get-AzVM'."
  ```
Please check the errors are call caused by the change of value piping and suppress them.

```
  D:\a\1\s\artifacts//StaticAnalysisResults\SignatureIssues.csv Errors
  "AssemblyFileName","ClassName","Target","Severity","ProblemId","Description","Remediation"
  "Az.Compute","Microsoft.Azure.Commands.Compute.StorageServices.AddAzureVhdCommand","Add-AzVhd","1","8420","Parameter set 'DefaultParameterSet' of cmdlet 'Add-AzVhd' contains at least one parameter with a position larger than four, which is discouraged.","Limit the number of positional parameters in a single parameter set to four or fewer."
```
Revert the change to SignatureIssues.csv

File: tools/StaticAnalysis/Exceptions/Az.Compute/BreakingChangeIssues.csv
Comment (Sandido) [reply tools/StaticAnalysis/Exceptions/Az.Compute/BreakingChangeIssues.csv]: @msJinLei , I updated it.

**Comment by Sandido** [comment]
> ```
>   D:\a\1\s\artifacts//StaticAnalysisResults\BreakingChangeIssues.csv Errors
>   "AssemblyFileName","ClassName","Target","Severity","ProblemId","Description","Remediation"
>   "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'DefaultParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'DefaultParamSet' back to cmdlet 'Get-AzVM'."
>   "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'GetVirtualMachineInResourceGroupParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'GetVirtualMachineInResourceGroupParamSet' back to cmdlet 'Get-AzVM'."
>   "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'ListLocationVirtualMachinesParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'ListLocationVirtualMachinesParamSet' back to cmdlet 'Get-AzVM'."
>   "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'ListNextLinkVirtualMachinesParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'ListNextLinkVirtualMachinesParamSet' back to cmdlet 'Get-AzVM'."
> ```
> 
> Please check the errors are call caused by the change of value piping and suppress them.
> 
> ```
>   D:\a\1\s\artifacts//StaticAnalysisResults\SignatureIssues.csv Errors
>   "AssemblyFileName","ClassName","Target","Severity","ProblemId","Description","Remediation"
>   "Az.Compute","Microsoft.Azure.Commands.Compute.StorageServices.AddAzureVhdCommand","Add-AzVhd","1","8420","Parameter set 'DefaultParameterSet' of cmdlet 'Add-AzVhd' contains at least one parameter with a position larger than four, which is discouraged.","Limit the number of positional parameters in a single parameter set to four or fewer."
> ```
> 
> Revert the change to SignatureIssues.csv



> ```
>   D:\a\1\s\artifacts//StaticAnalysisResults\BreakingChangeIssues.csv Errors
>   "AssemblyFileName","ClassName","Target","Severity","ProblemId","Description","Remediation"
>   "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'DefaultParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'DefaultParamSet' back to cmdlet 'Get-AzVM'."
>   "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'GetVirtualMachineInResourceGroupParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'GetVirtualMachineInResourceGroupParamSet' back to cmdlet 'Get-AzVM'."
>   "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'ListLocationVirtualMachinesParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'ListLocationVirtualMachinesParamSet' back to cmdlet 'Get-AzVM'."
>   "Az.Compute","Microsoft.Azure.Commands.Compute.GetAzureVMCommand","Get-AzVM","0","1050","The parameter set 'ListNextLinkVirtualMachinesParamSet' for cmdlet 'Get-AzVM' has been removed.","Add parameter set 'ListNextLinkVirtualMachinesParamSet' back to cmdlet 'Get-AzVM'."
> ```
> 
> Please check the errors are call caused by the change of value piping and suppress them.
> 
> ```
>   D:\a\1\s\artifacts//StaticAnalysisResults\SignatureIssues.csv Errors
>   "AssemblyFileName","ClassName","Target","Severity","ProblemId","Description","Remediation"
>   "Az.Compute","Microsoft.Azure.Commands.Compute.StorageServices.AddAzureVhdCommand","Add-AzVhd","1","8420","Parameter set 'DefaultParameterSet' of cmdlet 'Add-AzVhd' contains at least one parameter with a position larger than four, which is discouraged.","Limit the number of positional parameters in a single parameter set to four or fewer."
> ```
> 
> Revert the change to SignatureIssues.csv

I updated this.

**Comment by Sandido** [comment]
@msJinLei , The changes I made shouldn't have caused the breaking change and signature issue flags that occurred. I think this was an issue where I made edits to the cmdlet parameters (not additions) and that caused some extra tooling to run and detected edits to the cmdlets that occurred many years ago.

**Review by msJinLei — APPROVED** [review]

**Comment by msJinLei** [comment]
> @msJinLei , The changes I made shouldn't have caused the breaking change and signature issue flags that occurred. I think this was an issue where I made edits to the cmdlet parameters (not additions) and that caused some extra tooling to run and detected edits to the cmdlets that occurred many years ago.

Got approval by @dingmeng-xue offline. Will include this PR into release-2022-04-01

File: src/Compute/Compute/Common/ValidateBase64EncodedString.cs
Comment (Sandido) [inline src/Compute/Compute/Common/ValidateBase64EncodedString.cs:18]: I ended up adding this functional change because there seems to be a bug with the UserData parameter that it is sometimes registered as a Bound Parameter but with a value of Null. This is some extra defensive coding to avoid unnecessary validation run on an empty string.

Comment (Sandido) [reply src/Compute/Compute/Common/ValidateBase64EncodedString.cs:18]: Meaning this check will see an empty string or null as base64 encoded by default.

**Comment by Sandido** [comment]
@msJinLei , thank you for the timely merge!

**Pull request merged**
